modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/LogicalRelImplementor.java

```diff
@@ -1168,6 +1155,20 @@ private static boolean canFuseProjectionInto(RelNode rel) {
         return joinProjection;
     }
 
+    private long validateAndGetFetchOffsetParams(RexNode node, String op) {
+        SqlScalar<RowT, Number> sqlScalar = expressionFactory.scalar(node);
+        Number param = sqlScalar.get(ctx);
+
+        if (param instanceof BigDecimal) {
+            BigDecimal param0 = (BigDecimal) param;
+            if (param0.signum() == -1 || param0.compareTo(LIMIT_UPPER) > 0) {
+                throw new SqlException(Sql.STMT_VALIDATION_ERR, IgniteResource.INSTANCE.illegalFetchLimit(op).str());
+            }
+        }
```

Comment:
this covers only `BigDecimal`, but parameter may be of any numeric type. I would suggest to use `org.apache.ignite.internal.sql.engine.util.IgniteMath#convertToLongExact(java.lang.Number)` to cover overflow part, and then validate result of conversion to make sure it's not negative. Let's also improve test coverage

Reply:
done